### PR TITLE
TypeSpec CI Pipeline

### PR DESCRIPTION
**Pull request by mikeharder**

### CODE REVIEW

File: eng/pipelines/templates/steps/typespec-ci.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  type: string
+
+steps:
+- script: |
+    # Log commands before running
+    set -x
+
+    exitcode=0
+
+    cd ${{parameters.Folder}}
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/typespec-ci.yml:14]: Not sure if it matters much but you could also make this the working directory of the step.

File: eng/pipelines/templates/steps/typespec-ci.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    npx tsp compile .
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/typespec-ci.yml:15]: Do we need to look at the exit code?

**Review by weshaggard — APPROVED** [review]
Couple suggestions but otherwise looks reasonable.

File: eng/pipelines/typespec-ci.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    name: azsdk-pool-mms-ubuntu-2204-general
+    vmImage: ubuntu-22.04
+
+  steps:
+  - script: npm ci
+    displayName: npm ci
+  
+  - template: templates/steps/typespec-ci.yml
+    parameters:
+      Folder: specification/contosowidgetmanager/Contoso.WidgetManager
```
Comment (weshaggard) [inline eng/pipelines/typespec-ci.yml:13]: I know you are hardcoding for now, but we should definitely do a dynamic discovery of the projects eventually.

File: eng/pipelines/templates/steps/typespec-ci.yml (hunk 2 above)
Comment (mikeharder) [reply eng/pipelines/templates/steps/typespec-ci.yml:15]: Yes I need to fix error propagation

File: eng/pipelines/typespec-ci.yml (hunk 3 above)
Comment (mikeharder) [reply eng/pipelines/typespec-ci.yml:13]: Yes, I would like to discover projects dynamically, but I also think it's very nice to have one step per project, so you can easily see in the steps list which projects are failing.  So I will need to dynamically generate a matrix which is a little tricky (but Ben knows how).

File: eng/pipelines/templates/steps/typespec-ci.yml (hunk 1 above)
Comment (mikeharder) [reply eng/pipelines/templates/steps/typespec-ci.yml:14]: We go back to the repo root to run `git` commands, so I'd rather not use this as working directory.  Also, we should be able to use `tsp compile path/to/spec` from repo root once the bugs with that command are fixed.

File: eng/pipelines/templates/steps/typespec-ci.yml (hunk 2 above)
Comment (mikeharder) [reply eng/pipelines/templates/steps/typespec-ci.yml]: Should be fixed now

File: eng/pipelines/typespec-ci.yml (hunk 3 above)
Comment (weshaggard) [reply eng/pipelines/typespec-ci.yml:13]: I don't think I would do a dynamic matrix as that would run these all on their own agent which is probably not efficient. We would need to think more about how to run them in their own step from a dynamic list.

Comment (mikeharder) [reply eng/pipelines/typespec-ci.yml:13]: Yeah a matrix might not work here.  We can validate all the specs dynamically in a single script, but we'll need to make sure it's easy to see a summary of the results (which specs passed and failed), and carefully handle error propagation.

**Pull request merged**